### PR TITLE
basefw: Add handling of IPC4_DMA_CONTROL messages

### PR DESCRIPTION
**Pull request by tmleman**
This pull request introduces a handling of IPC4_DMA_CONTROL messages for the SSP DAI driver. 

Below is a summary of the commits included in this pull request:
- 64ce0b9b31b91b17f63cd435cc11d93fd3be28be Exposes the function to retrieve Zephyr DAI device structures, allowing other parts of the SOF codebase to access Zephyr native DAI drivers.
- 3801123335de58e0f5f2782f66beeffded1323c6: Implements handling for the IPC4_DMA_CONTROL message within the base firmware.

TODO before merge:

- [x] https://github.com/zephyrproject-rtos/zephyr/pull/73158
- [x] https://github.com/thesofproject/sof/pull/9338

### CODE REVIEW

File: src/audio/base_fw_intel.c
Comment (lyakh) [inline src/audio/base_fw_intel.c:335]: same here:
```
if (!dev) {
    tr_err(...);
    return -E...;
}
```

**Review by lyakh — COMMENTED** [review]
Thanks for addressing the comments! Formally the PR looks good to me now, although I still don't know why exactly this is needed - "improving the flexibility" doesn't tell me much, sorry :-) Would be nice to know what specific use-cases this is addressing.

**Review by lgirdwood — APPROVED** [review]

**Review by softwarecki — APPROVED** [review]

**Comment by lgirdwood** [comment]
@tmleman any update on ETA ? Do you have the Zephyr update now merged ?

**Review by iganakov — APPROVED** [review]

File: src/audio/base_fw_intel.c
Comment (iganakov) [inline src/audio/base_fw_intel.c:332]: I think this check should be removed. UAOL and DMIC interfaces will also implement DMA_CONTROL IPC to update dai configuration in runtime.

**Comment by lgirdwood** [comment]
@tmleman I guess we are blocked until after Zephyr merge Window ?

**Comment by tmleman** [comment]
> I guess we are blocked until after Zephyr merge Window ?

@lgirdwood Yes, this change is blocked until the release of Zephyr LTS.

File: src/audio/base_fw_intel.c
Comment (tmleman) [reply src/audio/base_fw_intel.c]: Thank you for pointing that out. The check for SSP-specific node IDs has been removed as the `dai_config_update` function will return `-ENOSYS` if a DAI driver does not implement the required operation. This ensures that the `IPC4_DMA_CONTROL` handling is appropriately generalized for all DAI types.

**Comment by lgirdwood** [comment]
> > I guess we are blocked until after Zephyr merge Window ?
> 
> @lgirdwood Yes, this change is blocked until the release of Zephyr LTS.

ok, tagged for v2.11 to remind everyone.

**Review by iganakov — CHANGES_REQUESTED** [review]
Please use IPC4 error codes

File: src/audio/base_fw.c
Comment (iganakov) [inline src/audio/base_fw.c:564]: I would rather pass `dma_control->config_data`  and `dma_control->config_length` as is here. Tlv data struct, which is typically used in `config_data` is based on DW, so It will be easier to handle it later.

**Review by iganakov — CHANGES_REQUESTED** [review]

**Comment by tmleman** [comment]
Rebase

**Review by lgirdwood — APPROVED** [review]
LGTM, but lets get alignment with @iganakov

File: src/audio/base_fw.c
Comment (kv2019i) [reply src/audio/base_fw.c:553]: @tmleman can you comment, is the config_length used or not?

**Review by kv2019i — COMMENTED** [review]
Looks good to me. If you can conclude the one discussion with @iganakov w.r.t. dma_control->config_length, this is good to go.

File: src/audio/base_fw_intel.c
Comment (tlissows) [reply src/audio/base_fw_intel.c:338]: This has already been corrected (IPC4 error codes used) but was later reverted (errno codes used again). Was this revert intended?

File: src/audio/base_fw.c
Comment (tlissows) [reply src/audio/base_fw.c:549]: This has already been corrected (IPC4 error codes used) but was later reverted (errno codes used again). Was this revert intended?

**Review by tlissows — CHANGES_REQUESTED** [review]

File: src/audio/base_fw.c
Comment (tmleman) [reply src/audio/base_fw.c]: I did a rebase on a machine that had the old code. My mistake, I've already corrected it.

File: src/audio/base_fw.c
Comment (tmleman) [reply src/audio/base_fw.c]: And from my perspective, it was correct (however, I focused only on SSP during the implementation). I looked at how it looks in the reference fw and changed the logic in the IF. Now the error is returned when `config_length` is greater than the actual data size. The condition you proposed will not work for SSP because in all cases I was able to check, this parameter is equal to zero, which is consistent with the description in the documentation.

File: src/audio/base_fw.c
Comment (tmleman) [reply src/audio/base_fw.c:563]: As I already wrote above, `config_length` has a value of 0 in the case of SSP and such logic will not work there.

File: src/audio/base_fw.c
Comment (tlissows) [inline src/audio/base_fw.c:555]: `data_size` is expressed in bytes while `dma_control->config_length` in dwords. Please take it into account in this comparison.

**Review by tlissows — APPROVED** [review]

**Review by iganakov — APPROVED** [review]
Looks good to me

**Comment by kv2019i** [comment]
Ready to go, but we need a clean pass of mandatory tests.

**Comment by kv2019i** [comment]
SOFCI TEST

**Pull request merged**
